File: Data Structures & Algorithms/top-k-elements-in-list/submission-2.py
from collections import Counter
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        # Count it with hashmap
        # Use max heap

        # 1. Count with hashmap
        # 2. save into a list as a (count, number) tuple
        # 3. Pop k times

# nums = [1,2,2,3,3,3], k = 2
        counter = Counter(nums) #{1: 1, 2: 2, 3: 3}

        counterArray = []
        for key, value in counter.items():
            counterArray.append((value, key))
        counterArray.sort() #[(1,1), (2,2), (3,3)]

        ans = []
        for i in range(k):
            counter, key = counterArray.pop()
            ans.append(key)
        return ans




    

        

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-2.js
class Solution {
    /**
     * @param {string} s
     * @param {number} k
     * @return {number}
     */
    characterReplacement(s, k) {
        // "AAABABB"
        //  ^    ^
        // count = {A:4, B:2}
        // res = 0
        //  l = 0
        //  r = 4
        // len = 6
        const count = {}
        let res = 0

        let l = 0
        for (let r = 0; r < s.length; r++){
            count[s[r]] = (count[s[r]] || 0 ) + 1

            while ((r-l+1) - Math.max(...Object.values(count)) > k){
                count[s[l]] -= 1
                l += 1
            }
            res = Math.max(r-l+1, res)
        }
        return res

    }
}
